cli/util/fs.rs

```diff
@@ -165,53 +165,101 @@ pub fn resolve_from_cwd(path: &Path) -> Result<PathBuf, AnyError> {
 
 /// Collects file paths that satisfy the given predicate, by recursively walking `files`.
 /// If the walker visits a path that is listed in `ignore`, it skips descending into the directory.
-pub fn collect_files<P>(
-  files: &[PathBuf],
-  ignore: &[PathBuf],
-  predicate: P,
-) -> Result<Vec<PathBuf>, AnyError>
-where
-  P: Fn(&Path) -> bool,
-{
-  let mut target_files = Vec::new();
-
-  // retain only the paths which exist and ignore the rest
-  let canonicalized_ignore: Vec<PathBuf> = ignore
-    .iter()
-    .filter_map(|i| canonicalize_path(i).ok())
-    .collect();
+pub struct FileCollector<TFilter: Fn(&Path) -> bool> {
+  canonicalized_ignore: Vec<PathBuf>,
+  file_filter: TFilter,
+  ignore_git_folder: bool,
+  ignore_node_modules: bool,
+}
 
-  for file in files {
-    for entry in WalkDir::new(file)
-      .into_iter()
-      .filter_entry(|e| {
-        canonicalize_path(e.path()).map_or(false, |c| {
-          !canonicalized_ignore.iter().any(|i| c.starts_with(i))
-        })
-      })
-      .filter_map(|e| match e {
-        Ok(e) if !e.file_type().is_dir() && predicate(e.path()) => Some(e),
-        _ => None,
-      })
-    {
-      target_files.push(canonicalize_path(entry.path())?)
+impl<TFilter: Fn(&Path) -> bool> FileCollector<TFilter> {
+  pub fn new(file_filter: TFilter) -> Self {
+    Self {
+      canonicalized_ignore: Default::default(),
+      file_filter,
+      ignore_git_folder: false,
+      ignore_node_modules: false,
     }
   }
+  pub fn add_ignore_paths(mut self, paths: &[PathBuf]) -> Self {
+    // retain only the paths which exist and ignore the rest
+    self
+      .canonicalized_ignore
+      .extend(paths.iter().filter_map(|i| canonicalize_path(i).ok()));
+    self
+  }
+
+  pub fn ignore_node_modules(mut self) -> Self {
+    self.ignore_node_modules = true;
+    self
+  }
+
+  pub fn ignore_git_folder(mut self) -> Self {
+    self.ignore_git_folder = true;
+    self
+  }
 
-  Ok(target_files)
+  pub fn collect_files(
+    &self,
+    files: &[PathBuf],
+  ) -> Result<Vec<PathBuf>, AnyError> {
+    let mut target_files = Vec::new();
+    for file in files {
+      if let Ok(file) = canonicalize_path(file) {
+        // use an iterator like this in order to minimize the number of file system operations
```

Comment:
Refactored this away from using `filter_entry` to reduce the `canonicalize_path` calls from two to one per path.